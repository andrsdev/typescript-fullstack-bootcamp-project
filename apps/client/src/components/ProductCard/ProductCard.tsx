import { Product } from '@repo/shared'
import { Link } from '@tanstack/react-router'
import { priceFormatter } from '../../utils'
interface ProductCardProps {
  product: Product
}

export const ProductCard = (props: ProductCardProps) => {
  const { product } = props
  return (
    <Link to={`/products/${product.id}`}>
      <div className="p-2">
        <div className="bg-white shadow-lg rounded-lg p-4 hover:shadow-xl transition-shadow duration-300">
          <img
            src={product.image}
            alt={product.name}
            className="w-full h-48 object-cover object-center rounded-lg"
          />
          <div className="mt-4">
            <div className="flex justify-between items-center">
              <p className="text-lg font-semibold text-gray-800 transition-colors duration-300 hover:text-indigo-600">
                {product.name}
              </p>
              <button className="bg-indigo-600 text-white text-sm font-bold py-1 px-4 rounded-full hover:bg-indigo-500 transition-colors duration-300">
                {priceFormatter(product.price)}
              </button>
            </div>
          </div>
        </div>
      </div>
    </Link>
  )
}
