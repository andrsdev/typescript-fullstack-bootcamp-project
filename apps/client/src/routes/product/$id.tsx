import { BaseResponseGeneric, ProductDtoResponse } from '@repo/schemas';
import { useQuery } from '@tanstack/react-query';
import { createFileRoute } from '@tanstack/react-router'
import { Button } from '../../components/Button';
import { useState } from 'react';
import { RadioGroup } from '../../components/RadioGroup';
import { RadioGroupItem } from '../../components/RadioGroupItem';


export const Route = createFileRoute('/product/$id')({
    component: ProductPage
})

function ProductPage() {
    const { id } = Route.useParams();
    const [color, setColor] = useState('black');
    const [selectedVariant, setSelectedVariant] = useState<string | null>(null);

    const { data, isLoading, error } = useQuery<BaseResponseGeneric<ProductDtoResponse>>({
        queryKey: ['products', id],
        queryFn: () => fetch(`http://localhost:5001/api/products/${id}`)
            .then((result) => result.json())
    })

    if (isLoading) {
        return <div>Loading...</div>;
    }

    if (error) {
        return <div>Error loading product</div>;
    }
    const product = data?.data;
    return (
        <div className="bg-gray-50 flex items-center justify-center p-4">
            <div className="bg-white rounded-lg shadow-lg p-6 max-w-3xl w-full">
                <div className="grid md:grid-cols-2 gap-6">
                    <div className="space-y-4">
                        {/* Product image */}
                        <div className="relative aspect-square bg-gray-100 rounded-lg overflow-hidden">
                            <img
                                src={product?.image || '/placeholder.svg?height=400&width=400'}
                                alt={`${product?.name}`}
                                className="object-cover w-full h-full"
                            />
                        </div>
                    </div>

                    {/* Product details */}
                    <div className="space-y-6">
                        <div>
                            <h1 className="text-3xl font-bold">{product?.name || 'Product Name'}</h1>
                            <div className="mt-2 inline-block bg-blue-500 text-white px-3 py-1 rounded-full text-sm font-semibold">
                                ${product?.price.toFixed(2) || '0.00'} USD
                            </div>
                        </div>

                        <div>
                            <h2 className="text-lg font-semibold mb-2">Editions</h2>
                            <RadioGroup value={color} onValueChange={setColor}>
                                <div className="flex space-x-2">
                                    <div className="flex items-center space-x-2">
                                        {product?.variants.map((variant) => (
                                            <div key={variant.id} className="flex items-center space-x-2">
                                                <RadioGroupItem
                                                    value={variant.name}
                                                    id={variant.id.toString()}
                                                    checked={selectedVariant === variant.name}
                                                    onChange={() => setSelectedVariant(variant.name)}
                                                />
                                            </div>
                                        ))}
                                    </div>
                                </div>
                            </RadioGroup>
                        </div>

                        <p className="text-gray-600">
                            {product?.description || 'Product description goes here.'}
                        </p>

                        <Button className="w-full">
                            Add To Cart
                        </Button>
                    </div>
                </div>
            </div>
        </div>
    );

}