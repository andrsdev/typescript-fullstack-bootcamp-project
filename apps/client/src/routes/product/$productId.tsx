import { createFileRoute } from '@tanstack/react-router'
import { ProductDetail } from '../../store/Pages/ProductDetail'

export const Route = createFileRoute('/product/$productId')({
  component: ProductPage,
})

function ProductPage() {
  const { productId } = Route.useParams()
  const id = Number(productId)

  return <ProductDetail id={id} />
}
