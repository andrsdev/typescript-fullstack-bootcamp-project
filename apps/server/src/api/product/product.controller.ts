import { Request, Response } from 'express'
import {
  createProduct,
  createProductCollection,
  createVariants,
  deleteProduct,
  getProductById,
  getProducts,
  getProductsByCollection,
  updateProduct,
} from './product.model'
import {
  Product,
  ProductParams,
  ProductQuery,
  UpdateProduct,
} from '../interfaces/product.interface'

export async function listProducts(
  req: Request<object, object, object, ProductQuery>,
  res: Response,
) {
  const sort = req.query.sort || 'desc'
  const filter = req.query.name || ''

  const result = await getProducts(sort, filter)

  return res.json({ products: result })
}

export async function productById(req: Request<ProductParams>, res: Response) {
  try {
    const productId = Number(req.params.id) || 0

    const result = await getProductById(productId)

    if (result === null) {
      return res
        .status(404)
        .send({ message: `Product with ID (${productId}) wasn't found.` })
    }
    return res.json({ result: [result] })
  } catch (e) {
    console.log((e as Error).message)
    return res.status(409).send({ error: (e as Error).message })
  }
}

export async function newProduct(req: Request<Product>, res: Response) {
  try {
    const product: Product = req.body

    const newProduct = await createProduct(product)
    const productId = newProduct.id

    const variants = product.variants
    await createVariants(productId, variants)

    const collectionId = Number(product.collection)
    await createProductCollection(productId, collectionId)

    return res.json({ product: newProduct })
  } catch (e) {
    console.log((e as Error).message)
    return res.status(409).send({ error: (e as Error).message })
  }
}

export async function modifyProduct(
  req: Request<UpdateProduct>,
  res: Response,
) {
  try {
    const productId = Number(req.params.id) || 0

    const product: UpdateProduct = req.body

    const updated = await updateProduct(productId, product)

    return res.json({ product: updated })
  } catch (e) {
    console.log((e as Error).message)
    return res.status(409).send({ error: (e as Error).message })
  }
}

export async function removeProduct(
  req: Request<ProductParams>,
  res: Response,
) {
  let productId = req.params.id || 0

  try {
    productId = Number(productId)
    const result = await deleteProduct(productId)
    if (result === null) {
      return res.status(404).send({
        message: `Cannot delete Product with ID (${productId}) because doesn't exist.`,
      })
    }
    return res.json({ result })
  } catch (e) {
    console.log((e as Error).message)
    return res.status(409).send({
      error: `Cannot delete Product with ID (${productId}) because doesn't exist.`,
    })
  }
}

export async function listProductsByCollection(
  req: Request<ProductParams, object, object, ProductQuery>,
  res: Response,
) {
  try {
    const collectionId = Number(req.params.id) || 0
    const sort = req.query.sort
    const filter = req.query.name || ''

    console.log(filter)

    const result = await getProductsByCollection(collectionId, sort, filter)

    if (result === null) {
      return res.status(404).send({
        message: `Products of collection (${collectionId}) wasn't found.`,
      })
    }

    return res.json({ products: result })
  } catch (e) {
    console.log((e as Error).message)
    return res.status(409).send({ error: (e as Error).message })
  }
}
