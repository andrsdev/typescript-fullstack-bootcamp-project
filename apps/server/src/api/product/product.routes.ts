import express, { Router } from 'express'
import {
  listProducts,
  listProductsByCollection,
  modifyProduct,
  newProduct,
  productById,
  removeProduct,
} from './product.controller'

const productRouter: Router = express.Router()

productRouter.get('/', listProducts)
productRouter.get('/:id', productById)
productRouter.post('/', newProduct)
productRouter.put('/:id', modifyProduct)
productRouter.delete('/:id', removeProduct)
productRouter.get('/collection/:id', listProductsByCollection)

export default productRouter
