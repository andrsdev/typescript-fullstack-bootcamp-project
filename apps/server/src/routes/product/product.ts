import express ,{ Express } from 'express'
import { prisma } from '../../lib/prismaClient';

export function productRouter(app: Express): void {
    const router = express.Router();
    app.use('/api/products', router);

    router.get('/', async (_req, res, next) => {
        try {
          const products = await prisma.product.findMany({
            include: { variants: true, collections: true },
          });
          res.status(200).json({statusCode: 200, data: products});
        } catch (error) {
          next({ statusCode: 500, message: 'Error fetching products '});
        }
      });
      
      router.get('/:id', async (req, res, next) => {
        try {
          const { id } = req.params;
          const product = await prisma.product.findUnique({
            where: { id: Number(id) },
            include: { variants: true, collections: true },
          });
          if (!product) {
            return next({ statusCode: 404, message: 'Product not found'});
          }
          res.status(200).json({statusCode: 200,data: product});
        } catch (error) {
          next({ statusCode: 500, message: 'Error fetching product'});
        }
      });
      
      router.post('/', async (req, res, next) => {
        try {
          const { name, description, price } = req.body;
          const newProduct = await prisma.product.create({
            data: { name, description, price },
          });
          res.status(201).json({statusCode:200, message:'The product was successfully created',data:newProduct});
        } catch (error) {
          next({ statusCode: 500, message: 'Fail during creation of product'});
        }
      });
      
      router.put('/:id', async (req, res, next) => {
        try {
          const { id } = req.params;
          const { name, description, price } = req.body;
          const updatedProduct = await prisma.product.update({
            where: { id: Number(id) },
            data: { name, description, price },
          });
          res.status(200).json({statusCode: 200, message: 'Success',data: updatedProduct});
        } catch (error) {
          next({ statusCode: 500, message: 'Product update failed'});
        }
      });
      
      router.delete('/:id', async (req, res, next) => {
        try {
          const { id } = req.params;
          await prisma.product.delete({
            where: { id: Number(id) },
          });
          res.status(204).json({ statusCode: 204 ,message: 'Product deleted' });
        } catch (error) {
          next({ statusCode: 500, message: 'Product deletion failed'});
        }
      });
    
}