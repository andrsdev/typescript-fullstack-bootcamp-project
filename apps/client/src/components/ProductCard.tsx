import { ProductCardProps } from '../interfaces/product.interface'
import { Link } from '@tanstack/react-router'

export const ProductCard: React.FC<ProductCardProps> = ({
  id,
  image,
  name,
  price,
}) => {
  return (
    <Link to={`/product/${id}`} key={id} className="product-card-link">
      <div className="product-card">
        <div className="card-image-container">
          <img
            src={
              'https://devchallengesyf.s3.us-east-2.amazonaws.com/products/' +
              image
            }
            alt={name}
          />
          <p className="hide">View Product</p>
        </div>
        <p className="card-title">{name}</p>
        <span className="price-badge">{'$' + price}</span>
      </div>
    </Link>
  )
}
