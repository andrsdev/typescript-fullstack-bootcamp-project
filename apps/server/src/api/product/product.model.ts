import { PrismaClient } from '@repo/db'
import { Variant } from '../interfaces/product.interface'

import { Product, UpdateProduct } from '../interfaces/product.interface'

const prisma = new PrismaClient()

export async function getProducts(
  sort: 'asc' | 'desc' = 'desc',
  filter: string,
) {
  return prisma.product.findMany({
    where: {
      name: {
        contains: filter,
        mode: 'insensitive',
      },
    },
    orderBy: [
      {
        price: sort,
      },
    ],
    include: {
      variants: true,
    },
  })
}

export async function getProductById(productId: number) {
  return prisma.product.findUnique({
    where: {
      id: productId,
    },
    include: {
      variants: true,
    },
  })
}

export async function createProduct(product: Product) {
  return prisma.product.create({
    data: {
      name: product.name,
      description: product.description || '',
      image: product.image || '',
      price: product.price,
      createdAt: new Date(product.createdAt).toISOString(),
      updatedAt: new Date(product.updatedAt).toISOString(),
    },
  })
}

export async function createVariants(productId: number, variants: Variant[]) {
  const variantsData = variants.map((v) => ({ ...v, productId }))

  return prisma.variant.createMany({
    data: [...variantsData],
  })
}

export async function createProductCollection(
  productId: number,
  collectionId: number,
) {
  return prisma.productCollection.create({
    data: {
      productId,
      collectionId,
    },
  })
}

export async function updateProduct(productId: number, product: UpdateProduct) {
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  const { id, ...data } = product

  return prisma.product.update({
    where: {
      id: productId,
    },
    data: data,
  })
}

export async function deleteProduct(productId: number) {
  return prisma.product.delete({
    where: {
      id: productId,
    },
    include: {
      variants: true,
      collections: true,
    },
  })
}

export async function getProductsByCollection(
  collectionId: number,
  sort: 'asc' | 'desc' = 'desc',
  nameFilter: string,
) {
  return await prisma.product.findMany({
    where: {
      collections: {
        some: {
          collectionId: collectionId,
        },
      },
      name: {
        contains: nameFilter,
        mode: 'insensitive',
      },
    },
    orderBy: {
      price: sort,
    },
    include: {
      variants: true, // Incluir variantes si es necesario
    },
  })
}
