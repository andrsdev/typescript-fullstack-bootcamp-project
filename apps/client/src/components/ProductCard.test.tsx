import { render } from '@testing-library/react'
import { ProductCard } from './ProductCard'
import { ProductCardProps } from '../interfaces/product.interface'

jest.mock('@tanstack/react-router', () => ({
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  Link: ({ children, ...props }) => <a {...props}>{children}</a>,
}))

describe('Tests on <CardProduct />', () => {
  const product: ProductCardProps = {
    id: 4,
    name: 'Dog Shampoo',
    image: 'dog-shampoo.png',
    price: 9.99,
  }

  test('Should match the snapshot', () => {
    const { container } = render(<ProductCard {...product} />)

    expect(container).toMatchSnapshot()
  })
})
