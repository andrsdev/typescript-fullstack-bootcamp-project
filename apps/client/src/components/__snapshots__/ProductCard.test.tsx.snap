// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests on <CardProduct /> Should match the snapshot 1`] = `
<div>
  <a
    class="product-card-link"
    to="/product/4"
  >
    <div
      class="product-card"
    >
      <div
        class="card-image-container"
      >
        <img
          alt=""
          src="https://devchallengesyf.s3.us-east-2.amazonaws.com/products/dog-shampoo.png"
        />
        <p
          class="hide"
        >
          View Product
        </p>
      </div>
      <p
        class="card-title"
      >
        Dog Shampoo
      </p>
      <span
        class="price-badge"
      >
        $9.99
      </span>
    </div>
  </a>
</div>
`;
